pkg/controller/nodelifecycle/node_lifecycle_controller.go

```diff
@@ -708,11 +708,22 @@ func (nc *Controller) doNoExecuteTaintingPass(ctx context.Context) {
 }
 
 func (nc *Controller) doEvictionPass(ctx context.Context) {
+	// Extract out the keys of the map in order to not hold
+	// the evictorLock for the entire function and hold it
+	// only when nescessary.
 	nc.evictorLock.Lock()
-	defer nc.evictorLock.Unlock()
+	zonePodEvictorKeys := make([]string, 0, len(nc.zonePodEvictor))
 	for k := range nc.zonePodEvictor {
+		zonePodEvictorKeys = append(zonePodEvictorKeys, k)
+	}
+	nc.evictorLock.Unlock()
+
+	for _, k := range zonePodEvictorKeys {
+		nc.evictorLock.Lock()
+		zonePodEvictionWorker := nc.zonePodEvictor[k]
```

Comment:
Shouldn't we check if it still exists in the map.

Also - is scheduler.RateLimitedTimedQueue safe to use without lock?

Reply:
> Shouldn't we check if it still exists in the map.

We should, good point. Thanks!

> is scheduler.RateLimitedTimedQueue safe to use without lock?

From what I understand, it is safe:

https://github.com/MadhavJivrajani/kubernetes/blob/db55f8b80d99ff8eeb14d8985f8461f4756f52b8/pkg/controller/nodelifecycle/scheduler/rate_limited_queue.go#L232-L235

To be sure, I modified `TestTry` to do:

```go
	wg := sync.WaitGroup{}
	wg.Add(10)
	for i := 0; i < 10; i++ {
		go func() {
			defer wg.Done()
			evictor.Try(func(value TimedValue) (bool, time.Duration) {
				deletedMap.Insert(value.Value)
				return true, 0
			})
		}()
	}
	wg.Wait()
```
and:
```
❯ go test -c -race
❯ stress ./scheduler.test -test.run ^TestTry$
5s: 311 runs so far, 0 failures
10s: 627 runs so far, 0 failures
15s: 881 runs so far, 0 failures
20s: 1114 runs so far, 0 failures
25s: 1346 runs so far, 0 failures
30s: 1579 runs so far, 0 failures
35s: 1817 runs so far, 0 failures
40s: 2047 runs so far, 0 failures
45s: 2282 runs so far, 0 failures
50s: 2518 runs so far, 0 failures
55s: 2748 runs so far, 0 failures
1m0s: 2953 runs so far, 0 failures
1m5s: 3190 runs so far, 0 failures
```

Reply:
> Edit: Now that I think about it, I'm not sure how a key would get deleted from the map. I don't see any logic around that, just additions happening here:

OK fair. It's for zones (and zones generally don't get added and removed), which means it's probably OK.

Can you please add a comment to mention say that it's safe?

Reply:
And I agree that zonePodEvictionWorker is thread-safe.